cpp/src/groupby/sort/functors.hpp

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/aggregation/result_cache.hpp>
+#include <cudf/detail/groupby/sort_helper.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <memory>
+
+namespace cudf {
+namespace groupby {
+namespace detail {
+/**
+ * @brief Functor to dispatch aggregation with
+ *
+ * This functor is to be used with `aggregation_dispatcher` to compute the
+ * appropriate aggregation. If the values on which to run the aggregation are
+ * unchanged, then this functor should be re-used. This is because it stores
+ * memoised sorted and/or grouped values and re-using will save on computation
+ * of these values.
+ */
+struct store_result_functor {
+  store_result_functor(size_type col_idx,
```

Comment:
I just saw you rename `store_result_functor` to `aggregrate_result_functor` in the `sort/aggregate.cpp` file. How do you name this functor? Should it be `aggregrate_result_functor` too?

Reply:
`store_result_functor` is base class. `aggregrate_result_functor` and `scan_result_functor` are derived classes.